src/client.rs

```diff
@@ -88,7 +153,106 @@ impl Record {
 
     // Return a field value.
     pub fn get(&self, key: &str) -> Option<&serde_json::Value> {
-        self.0.get(key)
+        self.value.get(key)
+    }
+
+    /// Return the attachment for this record, if any.
+    ///
+    /// Return values:
+    /// * `Some(Ok(_))` - There is an attachment, and it was successfully
+    ///   retrieved and converted to the requested type.
+    /// * `Some(Err(_))` - There is an attachment for this record, but it could
+    ///   not be retrieved or could not be converted to the requested type.
+    /// * `None` - There is no attachment for this record
+    ///
+    /// The `Some(Ok(_))` and `None` states are not expected to change. The
+    /// `Some(Err(_))` state may be temporary, and can be retried.
+    #[cfg(feature = "attachments")]
+    pub async fn get_attachment<T, E>(&mut self, client: &mut Client) -> Option<Result<T, ClientError>>
+    where T: TryFrom<Vec<u8>, Error = E>,
+        E: 'static + Send + Sync + std::error::Error
+    {
+        if let Attachment::Pending = self.attachment {
+            let maybe_attachment = self.value.as_object()
+                .and_then(|obj| obj.get("attachment"))
+                .map(|a| serde_json::from_value::<AttachmentMetadata>(a.clone()));
+
+            match maybe_attachment {
+                Some(Ok(meta)) => {
+                    self.attachment = Attachment::Some(meta);
+                }
+                Some(Err(err)) => {
+                    // serde_json::Error does not implement std::error::Error,
+                    // weirdly, so convert it to a string.
+                    return Some(Err(ClientError::AttachmentFormatError(
+                        anyhow!("Could not convert attachment to requested type: {}", err)
+                    )));
+                }
+                None => {
+                    self.attachment = Attachment::None;
+                }
+            }
+        }
+
+        match self.attachment {
+            Attachment::Some(ref metadata) => {
+                Some(Self::fetch_attachment::<T, E>(client, metadata).await)
+            },
+            Attachment::None => Option::None,
+            Attachment::Pending => unreachable!(),
+        }
+    }
+
+    #[cfg(feature = "attachments")]
+    async fn fetch_attachment<T, E>(
+        client: &mut Client,
+        metadata: &AttachmentMetadata
+    ) -> Result<T, ClientError>
+    where T: TryFrom<Vec<u8>, Error = E>,
+          E: 'static + Send + Sync + std::error::Error
+    {
+        let key = format!("attachment:{}", metadata.hash);
```

Comment:
Using the `hash` seems a reasonable approach to lookup the local cache :ok_hand:

Reply:
Do you think there is any need to delete old attachments? Merino isn't going to keep a cache that lives longer than a single node, so I'm not worried about it for that case.